sys/suit/coap.c

```diff
@@ -0,0 +1,508 @@
+                    DEBUG("nanocoap: maximum retries reached\n");
+                }
+                continue;
+            }
+            DEBUG("nanocoap: error receiving coap response, %d\n", (int)res);
+            break;
+        }
+        else {
+            if (coap_parse(pkt, (uint8_t *)buf, res) < 0) {
+                DEBUG("nanocoap: error parsing packet\n");
+                res = -EBADMSG;
+            }
+            break;
+        }
+    }
+
+    return res;
+}
+
+static int _fetch_block(coap_pkt_t *pkt, uint8_t *buf, sock_udp_t *sock, const char *path, coap_blksize_t blksize, size_t num)
+{
+    uint8_t *pktpos = buf;
+    pkt->hdr = (coap_hdr_t *)buf;
+
+    pktpos += coap_build_hdr(pkt->hdr, COAP_TYPE_CON, NULL, 0, COAP_METHOD_GET, num);
+    pktpos += coap_opt_put_uri_path(pktpos, 0, path);
+    pktpos += coap_put_option_block(pktpos, COAP_OPT_URI_PATH, num, blksize, 0, COAP_OPT_BLOCK2);
+
+    pkt->payload = pktpos;
+    pkt->payload_len = 0;
+
+    int res = _nanocoap_request(sock, pkt, 64 + (0x1 << (blksize + 4)));
+    if (res < 0) {
+        return res;
+    }
+
+    res = coap_get_code(pkt);
+    DEBUG("code=%i\n", res);
+    if (res != 205) {
+        return -res;
+    }
+
+    return 0;
+}
+
+int suit_coap_get_blockwise(sock_udp_ep_t *remote, const char *path,
+                               coap_blksize_t blksize,
+                               coap_blockwise_cb_t callback, void *arg)
+{
+    /* mmmmh dynamically sized array */
+    uint8_t buf[64 + (0x1 << (blksize + 4))];
+    sock_udp_ep_t local = SOCK_IPV6_EP_ANY;
+    coap_pkt_t pkt;
+
+    /* HACK: use random local port */
+    local.port = 0x8000 + (xtimer_now_usec() % 0XFFF);
+
+
+    sock_udp_t sock;
+    int res = sock_udp_create(&sock, &local, remote, 0);
+    if (res < 0) {
+        return res;
+    }
+
+
+    int more = 1;
+    size_t num = 0;
+    res = -1;
+    while (more == 1) {
+        DEBUG("fetching block %u\n", (unsigned)num);
+        res = _fetch_block(&pkt, buf, &sock, path, blksize, num);
+        DEBUG("res=%i\n", res);
+
+        if (!res) {
+            coap_block1_t block2;
+            coap_get_block2(&pkt, &block2);
+            more = block2.more;
+
+            if (callback(arg, block2.offset, pkt.payload, pkt.payload_len, more)) {
+                DEBUG("callback res != 0, aborting.\n");
+                res = -1;
+                goto out;
+            }
+        }
+        else {
+            DEBUG("error fetching block\n");
+            res = -1;
+            goto out;
+        }
+
+        num += 1;
+    }
+
+out:
+    sock_udp_close(&sock);
+    return res;
+}
+
+int suit_coap_get_blockwise_url(const char *url,
+                               coap_blksize_t blksize,
+                               coap_blockwise_cb_t callback, void *arg)
+{
+    char hostport[SOCK_HOSTPORT_MAXLEN];
+    char urlpath[SOCK_URLPATH_MAXLEN];
+    sock_udp_ep_t remote;
+
+    if (strncmp(url, "coap://", 7)) {
+        puts("no coap");
```

Comment:
Can we have better error messages (here and below) ? Maybe use LOG_ERROR ?

Reply:
fixed